crml/cennzx-spot/src/types.rs

```diff
@@ -13,91 +13,144 @@
 //
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
-//!
-//! CENNZX-SPOT Types
-//!
 use cennznet_primitives::types::AssetId;
-use core::convert::TryInto;
-use uint::construct_uint;
-construct_uint! {
-	/// 128-bit unsigned integer.
-	pub struct U128(2);
+use codec::{Decode, Encode, HasCompact};
+use core::{
+	convert::{From, Into, TryFrom},
+	marker::PhantomData,
+};
+use support::decl_error;
+
+pub use primitive_types::U256 as HighPrecisionUnsigned;
+pub use u128 as LowPrecisionUnsigned;
+
+/// A trait for values which hold an implicit scale factor that needs to be taken into account in calculations
+pub trait Scaled {
+	const SCALE: LowPrecisionUnsigned;
 }
 
-construct_uint! {
-	/// 256-bit unsigned integer.
-	pub struct U256(4);
+/// Per millionth of unit price
+#[derive(Debug, Clone, Copy)]
+pub enum PerMill {}
+impl Scaled for PerMill {
+	const SCALE: LowPrecisionUnsigned = 1_000_000;
 }
 
-use codec::{Compact, CompactAs, Decode, Encode, HasCompact};
+/// Per thousandth of unit price
+#[derive(Debug, Clone, Copy)]
```

Comment:
could u change this to `#[derive(Debug, Clone, Copy, PartialEq)]`

Reply:
Why we would need PartialEq here?

Reply:
So we can derive it for `FeeRate`